### PR TITLE
growpart: Add support for overprovisioning

### PR DESCRIPTION
**Pull request by dermotbradley**
Add option to 'growpart' to specify percentage of device that should be
left unallocated when growing partition. This is intended for consumer
SSDs and SD cards where the performance and/or lifetime of these devices
can be improved if some disk space (in addition to any the device "hides"
from users) is left unallocated.

### CODE REVIEW

File: bin/growpart
Comment (blackboxsw) [inline bin/growpart:392]: minor note in docs/logs I think over-provisioning is hyphenated

File: bin/growpart
Comment (blackboxsw) [inline bin/growpart:390]:
```suggestion
		# (for over-provisioning)
```

File: bin/growpart
Comment (blackboxsw) [inline bin/growpart:901]: In practice I don't think we can really accept 100 percent here. It's probably overkill, but I'm wondering if we should do a calculation of max_percent calculation based on the percentage represented by pt_end? We could alternately limit it to a hard-coded 90 percent or something? and document that acceptable range in the usage output.

**Review by blackboxsw — COMMENTED** [review]
quick pass. will actually exercise this in practice on a couple VMs to ensure I understand the implications.

File: bin/growpart
Comment (dermotbradley) [reply bin/growpart:901]: > In practice I don't think we can really accept 100 percent here

It doesn't, the logic is ">0 AND <100" so 1-99%.

I did think about having a smaller arbitary range, e.g. 5-10% or 5-15%, instead.

File: bin/growpart
Comment (dermotbradley) [reply bin/growpart:392]: I've seen both forms of spelling. It might be a British English vs American English difference.

File: bin/growpart
Comment (paride) [inline bin/growpart:393]: Worth checking that `max_end` is positive?

File: bin/growpart
Comment (paride) [reply bin/growpart:901]: I think this is good as it is, checking that `0 < x < 100`. It's the most basic sanity check, which I think it's mostly useful to catch malformed percentages (e.g. with decimals).

File: bin/growpart
Comment (paride) [reply bin/growpart:392]: As a non-native speaker I'll abstain here, let's just keep it consistent :) FWIW in the Debian codebase "overprovisioning" is way more common:

 - https://codesearch.debian.net/search?q=overprovisioning&literal=1
 - https://codesearch.debian.net/search?q=over-provisioning&literal=1

**Review by paride — COMMENTED** [review]
This is a nice addition, thanks! Before merging I think it would be good to have a simple test for it. Using https://github.com/canonical/cloud-utils/blob/main/test/test-growpart as a base this shouldn't be too much work: we grow the partition with `--free-percent` and then we check the number of blocks in `/proc/partitions`, somehow turning this:
```
echo "==== after ===="
grep "${lodev##*/}" /proc/partitions
```
into a check. No need to `resize2fs` IMO.

@dermotbradley would you be willing to add it to your PR?

File: bin/growpart
Comment (blackboxsw) [reply bin/growpart:390]: rejected

File: bin/growpart
Comment (blackboxsw) [reply bin/growpart:392]: +1 I'm good with overprovisioning. Sorry for delay there on a nit.

File: bin/growpart
Comment (blackboxsw) [reply bin/growpart:901]: Works for me thanks @paride. I think we we create an arbitrarly low MAX there'll be someone who is trying to test some funky config that we might want to allow at some point. All good as is, just wanted a little discussion on this. Thanks for the context @dermotbradley

Comment (dermotbradley) [reply bin/growpart:901]: @blackboxsw A realistic scenario when a large percentage figure might be used that I'm familiar with: using cloud-init on a Raspberry Pi which uses a SDcard for the OS. It is hard to find SDcards smaller than 4GB in size these days but the Alpine base disk image I build for RPIs is only 200-300MB or so in size so, when installed on a 4GB (or larger) SDcard even with data/logfile use I might wish to never partition more than 50% of the SDcard to (hopefully) extend its lifetime.

File: bin/growpart
Comment (dermotbradley) [reply bin/growpart:393]: Yupe, working on that currently

**Comment by dermotbradley** [comment]
> @dermotbradley would you be willing to add it to your PR?

Yes, working on that currently.

File: bin/growpart
Comment (dermotbradley) [reply bin/growpart]: I have rewritten the overprovisioning code and so I don't think this is now applicable.

**Comment by dermotbradley** [comment]
I have overhauled the overprovisioning code and added a testcase.

File: bin/growpart
Comment (paride) [inline bin/growpart:399]: Here we "floor" the number of free sectors; `$((sector_num*free_percent/100))` would be more precise, but I think what you did is fine, maybe even safer, even if we should still be afar from the bash arithmetic int overflow (64 bit signed).

One thing I was slightly worried about is partition alignment (see *align* in fdisk(8) and sfdisk(8)). At the moment, even without overprovisioning, growpart doesn't care about alignment, and just grows the partition to the last sector. This should be fine as aligning the end of the last partition is probably not critical, given that there are no more blocks the bad alignment can affect.

Should we care/worry about this when overprovisioning? Maybe we don't need to, trusting the SSD/SD firmware to do the right thing. Even if we do, then we should probably tackle the issue later in the script, for example by converting the amount to grow the partition by into a size in MiB, as sfdisk(8) recommends. I don't think think this has to be tackled in this PR.

This is more of a question than a review comment. @dermotbradley do you think above make sense, and do we agree it's not critical for landing this PR?

**Review by paride — COMMENTED** [review]
This looks great, thanks for adding the test! O just have one question (see inline comment).

File: test/test-growpart-overprovision
Comment (smoser) [inline test/test-growpart-overprovision:3]: I don't think there is really any reason that test of this function needs root.
tests that don't require root are much more safely run for a developer and thus more generally useful.

loop devices, but there isn't any real reason for that.
I think a test like test-growpart-fsimage would be just as good.

 * create a disk image
 * partition it
 * grow it
 * read the new partition table
 * verify expected.

File: bin/growpart
Comment (smoser) [reply bin/growpart:399]: to my knowledge, the *end* isn't really important for alignment except when it indicates (indirectly) the start of the next partition. when growpart grows to the full disk, there is no real value in end alignment and it can't easily affect beginning alignment. 

I guess if i were going to do things i'd go ahead and always end on a 1MiB-1 (or 4MiB-1) size but it doesn't seem terribly important.

File: test/test-growpart-overprovision
Comment (paride) [reply test/test-growpart-overprovision:3]: I agree, it would be nicer not to require root, especially as the test touches block devices (scary), however we're running it in CI, so I don't think it's worth blocking the PR from landing, wdyt?

Cc: @blackboxsw

Comment (dermotbradley) [reply test/test-growpart-overprovision:3]: @paride

Sorry for the lack of feedback - I have rewritten the test to not require root. Whilst doing I spotted what appeared to be some off-by-1 references to sectors in the growpart code, fixed those and some of the other tests broke, and I got a little lost in comparing sfdisk/sgdisk/parted output to get my mind straight regarding sectors :-)

Anyway I think I have it cracked now and this weekend expect to do another push to finalise things.

Comment (dermotbradley) [reply test/test-growpart-overprovision:3]: Ok. Additional changes pushed. That should be everything addressed.

File: bin/growpart
Comment (dermotbradley) [reply bin/growpart:401]: @paride I could change the order of the multiply/divide - indeed shellcheck recommends it the way you mentioned. Whichever way you prefer...

As for alignment, as @smoser pointed out growpart is only changing the end of the partition, not the start - in the majority of cases I'd assume its growing only the last (or only) partition on a device - that's certainly how I'm laying ouy my disk images, with rootfs as last partition. If its growing a partition other than the last one its still governed/restricted by the start of the next partition. So I don't believe alignment in this case is anything to worry about.

Comment (paride) [reply bin/growpart:401]: Re-thinking about integer overflows, I think it's safer to first divide and then multiply to avoid breaking 32 bit systems.

File: test/test-growpart-overprovision
Comment (paride) [reply test/test-growpart-overprovision:3]: @dermotbradley looks like the code in your PR still requires root, but you mentioned it should not require root now, so I'm a bit confused. Did you push your latest version?

Comment (dermotbradley) [reply test/test-growpart-overprovision:4]: @paride Ah, my mistake, I had pushed the wrong version of the file - corrected now.

**Review by paride — APPROVED** [review]
This LGTM, including the off-by-one fixes, thanks!

**Comment by dermotbradley** [comment]
Any chance to get this finally merged?

**Pull request merged**
